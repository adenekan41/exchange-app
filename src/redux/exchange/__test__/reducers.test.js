/* -------------------------- Internal Dependencies ------------------------- */
import ExchangeActions from '../types';
import exchangeReducer from '../reducers';

describe('Exchange Reducer', () => {
	let initialState;

	beforeEach(() => {
		initialState = {
			isLoading: false,
			from: 'USD',
			to: 'EUR',
			currentRate: 0.8342370902,
			wallet: {
				USD: {
					amount: 400,
				},
				GBP: {
					amount: 500,
				},
				EUR: {
					amount: 700,
				},
			},
		};
	});

	test('is correct', () => {
		const action = { type: 'dummy_action' };

		expect(exchangeReducer(undefined, action)).toEqual(initialState);
	});

	test('exchanges currency and update wallet and its balance', () => {
		const action = {
			type: ExchangeActions.EXCHANGE_CURRENCY,
			payload: {
				USD: {
					amount: 500,
				},
				GBP: {
					amount: 700,
				},
				EUR: {
					amount: 800,
				},
			},
		};
		initialState = {
			...initialState,
			wallet: {
				USD: {
					amount: 500,
				},
				GBP: {
					amount: 700,
				},
				EUR: {
					amount: 800,
				},
			},
		};

		expect(exchangeReducer(undefined, action)).toEqual(initialState);
		expect(exchangeReducer(undefined, action)).toMatchSnapshot();
	});
	test('start get currency and latest rates', () => {
		const action = {
			type: ExchangeActions.GET_CURRENCY,
		};
		initialState = {
			...initialState,
			isLoading: true,
		};

		expect(exchangeReducer(undefined, action)).toEqual(initialState);
		expect(exchangeReducer(undefined, action)).toMatchSnapshot();
	});
	test('successfully get currency and latest rates', () => {
		const action = {
			type: ExchangeActions.GET_CURRENCY_SUCCESS,
			payload: { data: 100 },
		};
		initialState = {
			...initialState,
			currentRate: 100,
			isLoading: false,
		};

		expect(exchangeReducer(undefined, action)).toEqual(initialState);
		expect(exchangeReducer(undefined, action)).toMatchSnapshot();
	});
	test('fail in getting currency and latest rates', () => {
		const action = {
			type: ExchangeActions.GET_CURRENCY_FAIL,
		};
		initialState = {
			...initialState,
			isLoading: false,
		};

		expect(exchangeReducer(undefined, action)).toEqual(initialState);
		expect(exchangeReducer(undefined, action)).toMatchSnapshot();
	});
	test('update FROM wallet', () => {
		const action = {
			type: ExchangeActions.UPDATE_FROM_CURRENT_WALLET,
			payload: 'USD',
		};
		initialState = {
			...initialState,
			from: 'USD',
		};

		expect(exchangeReducer(undefined, action)).toEqual(initialState);
		expect(exchangeReducer(undefined, action)).toMatchSnapshot();
	});
	test('update TO wallet', () => {
		const action = {
			type: ExchangeActions.UPDATE_TO_CURRENT_WALLET,
			payload: 'USD',
		};
		initialState = {
			...initialState,
			to: 'USD',
		};

		expect(exchangeReducer(undefined, action)).toEqual(initialState);
		expect(exchangeReducer(undefined, action)).toMatchSnapshot();
	});
});
