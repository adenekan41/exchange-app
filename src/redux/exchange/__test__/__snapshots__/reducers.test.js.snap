// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange Reducer exchanges currency and update wallet and its balance 1`] = `
Object {
  "currentRate": 0.8342370902,
  "from": "USD",
  "isLoading": false,
  "to": "EUR",
  "wallet": Object {
    "EUR": Object {
      "amount": 800,
    },
    "GBP": Object {
      "amount": 700,
    },
    "USD": Object {
      "amount": 500,
    },
  },
}
`;

exports[`Exchange Reducer fail in getting currency and latest rates 1`] = `
Object {
  "currentRate": 0.8342370902,
  "from": "USD",
  "isLoading": false,
  "to": "EUR",
  "wallet": Object {
    "EUR": Object {
      "amount": 700,
    },
    "GBP": Object {
      "amount": 500,
    },
    "USD": Object {
      "amount": 400,
    },
  },
}
`;

exports[`Exchange Reducer start get currency and latest rates 1`] = `
Object {
  "currentRate": 0.8342370902,
  "from": "USD",
  "isLoading": true,
  "to": "EUR",
  "wallet": Object {
    "EUR": Object {
      "amount": 700,
    },
    "GBP": Object {
      "amount": 500,
    },
    "USD": Object {
      "amount": 400,
    },
  },
}
`;

exports[`Exchange Reducer successfully get currency and latest rates 1`] = `
Object {
  "currentRate": 100,
  "from": "USD",
  "isLoading": false,
  "to": "EUR",
  "wallet": Object {
    "EUR": Object {
      "amount": 700,
    },
    "GBP": Object {
      "amount": 500,
    },
    "USD": Object {
      "amount": 400,
    },
  },
}
`;

exports[`Exchange Reducer update FROM wallet 1`] = `
Object {
  "currentRate": 0.8342370902,
  "from": "USD",
  "isLoading": false,
  "to": "EUR",
  "wallet": Object {
    "EUR": Object {
      "amount": 700,
    },
    "GBP": Object {
      "amount": 500,
    },
    "USD": Object {
      "amount": 400,
    },
  },
}
`;

exports[`Exchange Reducer update TO wallet 1`] = `
Object {
  "currentRate": 0.8342370902,
  "from": "USD",
  "isLoading": false,
  "to": "USD",
  "wallet": Object {
    "EUR": Object {
      "amount": 700,
    },
    "GBP": Object {
      "amount": 500,
    },
    "USD": Object {
      "amount": 400,
    },
  },
}
`;
